src/source-af-xdp.c

```diff
@@ -0,0 +1,944 @@
+
+    /* Busy polling configuration */
+    ptv->xsk.enable_busy_poll = afxdpconfig->enable_busy_poll;
+    ptv->xsk.busy_poll_budget = afxdpconfig->busy_poll_budget;
+    ptv->xsk.busy_poll_time = afxdpconfig->busy_poll_time;
+    ptv->gro_flush_timeout = afxdpconfig->gro_flush_timeout;
+    ptv->napi_defer_hard_irqs = afxdpconfig->napi_defer_hard_irqs;
+
+    /* Stats registration */
+    ptv->capture_afxdp_packets = StatsRegisterCounter("capture.afxdp_packets", ptv->tv);
+    ptv->capture_kernel_drops = StatsRegisterCounter("capture.kernel_drops", ptv->tv);
+    ptv->capture_afxdp_poll = StatsRegisterCounter("capture.afxdp.poll", ptv->tv);
+    ptv->capture_afxdp_poll_timeout = StatsRegisterCounter("capture.afxdp.poll_timeout", ptv->tv);
+    ptv->capture_afxdp_poll_failed = StatsRegisterCounter("capture.afxdp.poll_failed", ptv->tv);
+    ptv->capture_afxdp_empty_reads = StatsRegisterCounter("capture.afxdp.empty_reads", ptv->tv);
+    ptv->capture_afxdp_failed_reads = StatsRegisterCounter("capture.afxdp.failed_reads", ptv->tv);
+
+    *data = (void *)ptv;
+    afxdpconfig->DerefFunc(afxdpconfig);
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode AcquireBuffer(AFXDPThreadVars *ptv)
+{
+    int mmap_flags = MAP_PRIVATE | MAP_ANONYMOUS | ptv->umem.mmap_alignment_flag;
+    ptv->umem.buf = mmap(NULL, MEM_BYTES, PROT_READ | PROT_WRITE, mmap_flags, -1, 0);
+
+    if (ptv->umem.buf == MAP_FAILED) {
+        SCLogError(SC_ERR_MEM_ALLOC, "mmap: failed to acquire memory");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode ConfigureXSKUmem(AFXDPThreadVars *ptv)
+{
+    if (xsk_umem__create(&ptv->umem.umem, ptv->umem.buf, MEM_BYTES, &ptv->umem.fq, &ptv->umem.cq,
+                &ptv->umem.cfg)) {
+        SCLogError(SC_ERR_AFXDP_CREATE, "failed to create umem: %s", strerror(errno));
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode InitFillRing(AFXDPThreadVars *ptv, const uint32_t cnt)
+{
+    uint32_t idx_fq = 0;
+
+    uint32_t ret = xsk_ring_prod__reserve(&ptv->umem.fq, cnt, &idx_fq);
+    if (ret != cnt) {
+        SCLogError(SC_ERR_AFXDP_INIT, "Failed to initialise the fill ring.");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    for (uint32_t i = 0; i < cnt; i++) {
+        *xsk_ring_prod__fill_addr(&ptv->umem.fq, idx_fq++) = i * FRAME_SIZE;
+    }
+
+    xsk_ring_prod__submit(&ptv->umem.fq, cnt);
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/**
+ * \brief Linux knobs are tuned to enable a NAPI polling context
+ *
+ * \param tv pointer to AFXDPThreadVars
+ */
+static TmEcode WriteLinuxTunables(AFXDPThreadVars *ptv)
+{
+    char fname[DEFAULT_FILENAME_SIZE];
+
+    if (snprintf(fname, DEFAULT_FILENAME_SIZE, "/sys/class/net/%s/gro_flush_timeout", ptv->iface) <
+            0) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    FILE *fp = fopen(fname, "w");
+    if (fp == NULL) {
+        SCLogWarning(SC_ERR_FOPEN, "Could not open file: %s", fname);
```

Comment:
are these really warnings, or are they acting as errors? If the latter, please make them `SCLogError`